File: cpp/Binary_Tree_Level_Order_Traversal_II.cc
/*
Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).

For example:
Given binary tree {3,9,20,#,#,15,7},
    3
   / \
  9  20
    /  \
   15   7
return its bottom-up level order traversal as:
[
  [15,7],
  [9,20],
  [3]
]
*/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution 
{
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) 
    {
        vector<vector<int> > result;
        if (root == NULL)
            return result;
            
        queue<TreeNode*> current, next;
        vector<int> level;
        
        current.push(root);
        while (!current.empty())
        {
            while (!current.empty())
            {
                TreeNode* node = current.front();
                current.pop();
                
                level.push_back(node->val);
                
                if (node->left != NULL)
                    next.push(node->left);
                    
                if (node->right != NULL)
                    next.push(node->right);
            }
            
            result.push_back(level);
            level.clear();
            swap(next, current);
        }
        
        reverse(result.begin(), result.end()); 
        
        return result;
    }
};

File: go/binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.go
package binary_tree_level_order_traversal_ii

import (
	"../util"
)

/*
107. 二叉树的层次遍历 II

给定一个二叉树，返回其节点值自底向上的层次遍历。 （即按从叶子节点所在层到根节点所在的层，逐层从左向右遍历）

例如：
给定二叉树 [3,9,20,null,null,15,7],

    3
   / \
  9  20
    /  \
   15   7
返回其自底向上的层次遍历为：

[
  [15,7],
  [9,20],
  [3]
]
*/

type TreeNode = util.TreeNode

func levelOrderBottom(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	result := make([][]int, 0)
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) > 0 {
		level := make([]int, 0)

		l := len(queue)
		for i := 0; i < l; i++ {
			cur := queue[0]
			queue = queue[1:]
			level = append(level, cur.Val)

			if cur.Left != nil {
				queue = append(queue, cur.Left)
			}
			if cur.Right != nil {
				queue = append(queue, cur.Right)
			}
		}
		result = append([][]int{level}, result...)
	}

	return result
}
